<!DOCTYPE html>
<html>
<head><title>R: An Efficient Swiss Army Knife for Population Genomic Analyses</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> An Efficient Swiss Army Knife for Population Genomic Analyses
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;PopGenome&rsquo; version 2.7.7</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<p style="text-align: center;">
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#I">I</a>
<a href="#J">J</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#U">U</a>
<a href="#V">V</a>
<a href="#W">W</a>
</p>


<h2><a id="A">-- A --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">Achaz.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="Achaz.stats.html">Achaz.stats-method</a></td>
<td>Achaz statistic</td></tr>
<tr><td style="width: 25%;"><a href="Achaz.stats.html">Achaz.stats-methods</a></td>
<td>Achaz statistic</td></tr>
</table>

<h2><a id="B">-- B --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="BayeScanR.html">BayeScanR</a></td>
<td>An R implementation of BayeScan (Foll &amp; Gagiotti 2008)</td></tr>
</table>

<h2><a id="C">-- C --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">calc.fixed.shared</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="calc.fixed.shared-methods.html">calc.fixed.shared-method</a></td>
<td>Fixed and shared polymorphisms</td></tr>
<tr><td style="width: 25%;"><a href="calc.fixed.shared-methods.html">calc.fixed.shared-methods</a></td>
<td>Fixed and shared polymorphisms</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">calc.R2</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="calc.R2-methods.html">calc.R2-method</a></td>
<td>Linkage statistics (R2, P-value, Distance)</td></tr>
<tr><td style="width: 25%;"><a href="calc.R2-methods.html">calc.R2-methods</a></td>
<td>Linkage statistics (R2, P-value, Distance)</td></tr>
<tr><td style="width: 25%;"><a href="codontable.html">codontable</a></td>
<td>Prints the codon table which is used in the PopGenome framework</td></tr>
<tr><td style="width: 25%;"><a href="concatenate.classes.html">concatenate.classes</a></td>
<td>Concatenate GENOME classes</td></tr>
<tr><td style="width: 25%;"><a href="concatenate.regions.html">concatenate.regions</a></td>
<td>Concatenate regions</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">count.unknowns</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="count.unknowns-methods.html">count.unknowns-method</a></td>
<td>Calculate missing nucleotide frequencies</td></tr>
<tr><td style="width: 25%;"><a href="count.unknowns-methods.html">count.unknowns-methods</a></td>
<td>Calculate missing nucleotide frequencies</td></tr>
<tr><td style="width: 25%;"><a href="create.PopGenome.method.html">create.PopGenome.method</a></td>
<td>Integration of own functions into the PopGenome-framework</td></tr>
</table>

<h2><a id="D">-- D --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">detail.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="detail.stats-methods.html">detail.stats-method</a></td>
<td>Several statistics</td></tr>
<tr><td style="width: 25%;"><a href="detail.stats-methods.html">detail.stats-methods</a></td>
<td>Several statistics</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">diversity.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="diversity.stats-methods.html">diversity.stats-method</a></td>
<td>Diversities</td></tr>
<tr><td style="width: 25%;"><a href="diversity.stats-methods.html">diversity.stats-methods</a></td>
<td>Diversities</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">diversity.stats.between</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="diversity.stats.between-methods.html">diversity.stats.between-method</a></td>
<td>Diversities</td></tr>
<tr><td style="width: 25%;"><a href="diversity.stats.between-methods.html">diversity.stats.between-methods</a></td>
<td>Diversities</td></tr>
</table>

<h2><a id="F">-- F --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="fasta_file.html">fasta_file</a></td>
<td>FASTA file (subdirectory "data")</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">F_ST.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats-methods.html">F_ST.stats-method</a></td>
<td>Fixation Index</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats-methods.html">F_ST.stats-methods</a></td>
<td>Fixation Index</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">F_ST.stats.2</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats.2-methods.html">F_ST.stats.2-method</a></td>
<td>Fixation Index (2)</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats.2-methods.html">F_ST.stats.2-methods</a></td>
<td>Fixation Index (2)</td></tr>
</table>

<h2><a id="G">-- G --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">GENOME-class</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.biallelic.matrix</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="get.biallelic.matrix-methods.html">get.biallelic.matrix-method</a></td>
<td>Get the biallelic matrix</td></tr>
<tr><td style="width: 25%;"><a href="get.biallelic.matrix-methods.html">get.biallelic.matrix-methods</a></td>
<td>Get the biallelic matrix</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.codons</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="get.codons-methods.html">get.codons-method</a></td>
<td>Detailed information about the nature of codon changes</td></tr>
<tr><td style="width: 25%;"><a href="get.codons-methods.html">get.codons-methods</a></td>
<td>Detailed information about the nature of codon changes</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.detail</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="detail.stats-methods.html">get.detail-method</a></td>
<td>Several statistics</td></tr>
<tr><td style="width: 25%;"><a href="detail.stats-methods.html">get.detail-methods</a></td>
<td>Several statistics</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.diversity</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats-methods.html">get.diversity-method</a></td>
<td>Fixation Index</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats-methods.html">get.diversity-methods</a></td>
<td>Fixation Index</td></tr>
<tr><td style="width: 25%;"><a href="get.feature.names.html">get.feature.names</a></td>
<td>Feature informations and GFF-attributes</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.F_ST</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats-methods.html">get.F_ST-method</a></td>
<td>Fixation Index</td></tr>
<tr><td style="width: 25%;"><a href="F_ST.stats-methods.html">get.F_ST-methods</a></td>
<td>Fixation Index</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.individuals</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="get.individuals-methods.html">get.individuals-method</a></td>
<td>Print the names/IDs of individuals</td></tr>
<tr><td style="width: 25%;"><a href="get.individuals-methods.html">get.individuals-methods</a></td>
<td>Print the names/IDs of individuals</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.linkage</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="linkage.stats-methods.html">get.linkage-method</a></td>
<td>Linkage Disequilibrium</td></tr>
<tr><td style="width: 25%;"><a href="linkage.stats-methods.html">get.linkage-methods</a></td>
<td>Linkage Disequilibrium</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.MKT</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="MKT-methods.html">get.MKT-method</a></td>
<td>McDonald-Kreitman Test (McDonald &amp; Kreitman 1991)</td></tr>
<tr><td style="width: 25%;"><a href="MKT-methods.html">get.MKT-methods</a></td>
<td>McDonald-Kreitman Test (McDonald &amp; Kreitman 1991)</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.neutrality</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="neutrality.stats-methods.html">get.neutrality-method</a></td>
<td>Neutrality Statistics</td></tr>
<tr><td style="width: 25%;"><a href="neutrality.stats-methods.html">get.neutrality-methods</a></td>
<td>Neutrality Statistics</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.recomb</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="recomb.stats-methods.html">get.recomb-method</a></td>
<td>Recombination statistics</td></tr>
<tr><td style="width: 25%;"><a href="recomb.stats-methods.html">get.recomb-methods</a></td>
<td>Recombination statistics</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.status</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="get.status-methods.html">get.status-method</a></td>
<td>State of calculations</td></tr>
<tr><td style="width: 25%;"><a href="get.status-methods.html">get.status-methods</a></td>
<td>State of calculations</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.sum.data</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="readData.html">get.sum.data-method</a></td>
<td>Read alignments and calculate summary data</td></tr>
<tr><td style="width: 25%;"><a href="readData.html">get.sum.data-methods</a></td>
<td>Read alignments and calculate summary data</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">get.sweeps</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="sweeps.stats-methods.html">get.sweeps-method</a></td>
<td>Selective Sweeps</td></tr>
<tr><td style="width: 25%;"><a href="sweeps.stats-methods.html">get.sweeps-methods</a></td>
<td>Selective Sweeps</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">getBayes</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="getBayes-methods.html">getBayes-method</a></td>
<td>Get values for BayeScanR</td></tr>
<tr><td style="width: 25%;"><a href="getBayes-methods.html">getBayes-methods</a></td>
<td>Get values for BayeScanR</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">getMS-method</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="get_gff_info.html">get_gff_info</a></td>
<td>Annotation info</td></tr>
<tr><td style="width: 25%;"><a href="gff_file.html">gff_file</a></td>
<td>GFF file (subdirectory "data")</td></tr>
<tr><td style="width: 25%;"><a href="GFF_split_into_scaffolds.html">GFF_split_into_scaffolds</a></td>
<td>Split a GFF file into multiple scaffold-GFFs</td></tr>
</table>

<h2><a id="I">-- I --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">introgression.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="introgression.stats.html">introgression.stats-method</a></td>
<td>Introgression statistics</td></tr>
<tr><td style="width: 25%;"><a href="introgression.stats.html">introgression.stats-methods</a></td>
<td>Introgression statistics</td></tr>
</table>

<h2><a id="J">-- J --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">jack.knife.transform</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="jack.knife.transform-methods.html">jack.knife.transform-method</a></td>
<td>Jacknife Transformation</td></tr>
<tr><td style="width: 25%;"><a href="jack.knife.transform-methods.html">jack.knife.transform-methods</a></td>
<td>Jacknife Transformation</td></tr>
</table>

<h2><a id="L">-- L --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">linkage.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="linkage.stats-methods.html">linkage.stats-method</a></td>
<td>Linkage Disequilibrium</td></tr>
<tr><td style="width: 25%;"><a href="linkage.stats-methods.html">linkage.stats-methods</a></td>
<td>Linkage Disequilibrium</td></tr>
<tr><td style="width: 25%;"><a href="load.session.html">load.session</a></td>
<td>Loading a PopGenome session</td></tr>
</table>

<h2><a id="M">-- M --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">MKT</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="MKT-methods.html">MKT-method</a></td>
<td>McDonald-Kreitman Test (McDonald &amp; Kreitman 1991)</td></tr>
<tr><td style="width: 25%;"><a href="MKT-methods.html">MKT-methods</a></td>
<td>McDonald-Kreitman Test (McDonald &amp; Kreitman 1991)</td></tr>
<tr><td style="width: 25%;"><a href="MS.html">MS</a></td>
<td>Coalescent simulation with or without selection</td></tr>
<tr><td style="width: 25%;"><a href="MS_getStats.html">MS_getStats</a></td>
<td>Get the simulated MS/MSMS statistics</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">mult.linkage.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="mult.linkage.stats.html">mult.linkage.stats-method</a></td>
<td>Multilocus linkage statistics</td></tr>
<tr><td style="width: 25%;"><a href="mult.linkage.stats.html">mult.linkage.stats-methods</a></td>
<td>Multilocus linkage statistics</td></tr>
</table>

<h2><a id="N">-- N --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">neutrality.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="neutrality.stats-methods.html">neutrality.stats-method</a></td>
<td>Neutrality Statistics</td></tr>
<tr><td style="width: 25%;"><a href="neutrality.stats-methods.html">neutrality.stats-methods</a></td>
<td>Neutrality Statistics</td></tr>
</table>

<h2><a id="P">-- P --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">PG_plot.biallelic.matrix</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="PG_plot.biallelic.matrix.html">PG_plot.biallelic.matrix-method</a></td>
<td>Plot the biallelic matrix</td></tr>
<tr><td style="width: 25%;"><a href="PG_plot.biallelic.matrix.html">PG_plot.biallelic.matrix-methods</a></td>
<td>Plot the biallelic matrix</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">popFSTN-method</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="PopGenome.html">PopGenome</a></td>
<td>PopGenome</td></tr>
<tr><td style="width: 25%;"><a href="PopGplot.html">PopGplot</a></td>
<td>Smoothed line-plot for multiple populations</td></tr>
</table>

<h2><a id="R">-- R --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="read.big.fasta.html">read.big.fasta</a></td>
<td>Reading large FASTA alignments</td></tr>
<tr><td style="width: 25%;"><a href="readData.html">readData</a></td>
<td>Read alignments and calculate summary data</td></tr>
<tr><td style="width: 25%;"><a href="readHapMap.html">readHapMap</a></td>
<td>Read SNP data from the HapMap consortium</td></tr>
<tr><td style="width: 25%;"><a href="readMS.html">readMS</a></td>
<td>Read output data from MS and MSMS</td></tr>
<tr><td style="width: 25%;"><a href="readSNP.html">readSNP</a></td>
<td>Read data in .SNP format</td></tr>
<tr><td style="width: 25%;"><a href="readVCF.html">readVCF</a></td>
<td>Read SNP data in tabixed VCF format</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">recomb.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="recomb.stats-methods.html">recomb.stats-method</a></td>
<td>Recombination statistics</td></tr>
<tr><td style="width: 25%;"><a href="recomb.stats-methods.html">recomb.stats-methods</a></td>
<td>Recombination statistics</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">region.as.fasta</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="region.as.fasta.html">region.as.fasta-method</a></td>
<td>Extract a region and write it to a FASTA file</td></tr>
<tr><td style="width: 25%;"><a href="region.as.fasta.html">region.as.fasta-methods</a></td>
<td>Extract a region and write it to a FASTA file</td></tr>
</table>

<h2><a id="S">-- S --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="save.session.html">save.session</a></td>
<td>Save the '"GENOME"' object of a PopGenome session</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">set.filter</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="set.filter-methods.html">set.filter-method</a></td>
<td>Setting filter to the analysis</td></tr>
<tr><td style="width: 25%;"><a href="set.filter-methods.html">set.filter-methods</a></td>
<td>Setting filter to the analysis</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">set.outgroup</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="set.outgroup-methods.html">set.outgroup-method</a></td>
<td>Define an outgroup</td></tr>
<tr><td style="width: 25%;"><a href="set.outgroup-methods.html">set.outgroup-methods</a></td>
<td>Define an outgroup</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">set.populations</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="set.populations-methods.html">set.populations-method</a></td>
<td>Define populations</td></tr>
<tr><td style="width: 25%;"><a href="set.populations-methods.html">set.populations-methods</a></td>
<td>Define populations</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">set.ref.positions</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="set.ref.positions.html">set.ref.positions-method</a></td>
<td>Set reference positions for SNP data</td></tr>
<tr><td style="width: 25%;"><a href="set.ref.positions.html">set.ref.positions-methods</a></td>
<td>Set reference positions for SNP data</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">set.synnonsyn</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="set.synnonsyn-methods.html">set.synnonsyn-method</a></td>
<td>Set synonymous positions for SNP data</td></tr>
<tr><td style="width: 25%;"><a href="set.synnonsyn-methods.html">set.synnonsyn-methods</a></td>
<td>Set synonymous positions for SNP data</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">show-method</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">show.slots</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="show.slots-methods.html">show.slots-method</a></td>
<td>Show Slots of class GENOME</td></tr>
<tr><td style="width: 25%;"><a href="show.slots-methods.html">show.slots-methods</a></td>
<td>Show Slots of class GENOME</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">sliding.window.transform</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="sliding.window.transform-methods.html">sliding.window.transform-method</a></td>
<td>Sliding Window Transformation</td></tr>
<tr><td style="width: 25%;"><a href="sliding.window.transform-methods.html">sliding.window.transform-methods</a></td>
<td>Sliding Window Transformation</td></tr>
<tr><td style="width: 25%;"><a href="snp_file.html">snp_file</a></td>
<td>.SNP file (variant call data from 1001 Arabidopsis Genomes project)</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">splitting.data</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="splitting.data-methods.html">splitting.data-method</a></td>
<td>Split data into subsites</td></tr>
<tr><td style="width: 25%;"><a href="splitting.data-methods.html">splitting.data-methods</a></td>
<td>Split data into subsites</td></tr>
<tr><td style="width: 25%;"><a href="split_data_into_GFF_attributes.html">split_data_into_GFF_attributes</a></td>
<td>Split the data into GFF attributes</td></tr>
<tr><td style="width: 25%;"><a href="split_data_into_GFF_features.html">split_data_into_GFF_features</a></td>
<td>Split the data into GFF features</td></tr>
<tr><td style="width: 25%;"><a href="GENOME-class.html">sweeps.stats</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="sweeps.stats-methods.html">sweeps.stats-method</a></td>
<td>Selective Sweeps</td></tr>
<tr><td style="width: 25%;"><a href="sweeps.stats-methods.html">sweeps.stats-methods</a></td>
<td>Selective Sweeps</td></tr>
</table>

<h2><a id="T">-- T --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="test.params-class.html">test.params</a></td>
<td>Set parameters for coalescent simulations with Hudson's MS and Ewing's MSMS.</td></tr>
<tr><td style="width: 25%;"><a href="test.params-class.html">test.params-class</a></td>
<td>Set parameters for coalescent simulations with Hudson's MS and Ewing's MSMS.</td></tr>
</table>

<h2><a id="U">-- U --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">usage-method</a></td>
<td>Class "GENOME"</td></tr>
</table>

<h2><a id="V">-- V --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="vcf_file.html">vcf_file</a></td>
<td>VCF file (subdirectory "data")</td></tr>
<tr><td style="width: 25%;"><a href="VCF_split_into_scaffolds.html">VCF_split_into_scaffolds</a></td>
<td>Split a VCF file into multiple scaffold-VCFs</td></tr>
</table>

<h2><a id="W">-- W --</a></h2>

<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="GENOME-class.html">weighted.jackknife</a></td>
<td>Class "GENOME"</td></tr>
<tr><td style="width: 25%;"><a href="weighted.jackknife.html">weighted.jackknife-method</a></td>
<td>Weighted Jackknife</td></tr>
<tr><td style="width: 25%;"><a href="weighted.jackknife.html">weighted.jackknife-methods</a></td>
<td>Weighted Jackknife</td></tr>
<tr><td style="width: 25%;"><a href="Whop_readVCF.html">Whop_readVCF</a></td>
<td>Reading tabixed VCF files (an interface to WhopGenome)</td></tr>
</table>
</div></body></html>
